### PR TITLE
fix CP `n_tokens_seen` overcounting

### LINKED ISSUE #2983: n_tokens_seen metric inflated by context_parallel_degree when CP is enabled
### Bug description

When context parallelism is enabled, the `n_tokens_seen` metric reported in logs/wandb is inflated by `cp_degree`. This makes it impossible to compare training curves (e.g., loss vs tokens seen) across runs with different CP configurations.

`tps` and `mfu` are correctly normalized via `non_data_parallel_size`, so this only affects `n_tokens_seen`.

## Possible root cause

In `trainer.py` `batch_generator`, `labels.numel()` is counted before `context_parallel` chunks the labels:

https://github.com/pytorch/torchtitan/blob/main/torchtitan/trainer.py#L542-L544

```python
ntokens_batch = labels.numel()          # full seq_len, before CP split
self.ntokens_seen += ntokens_batch
self.metrics_processor.ntokens_since_last_log += ntokens_batch
```

Then `n_tokens_seen` is aggregated via `dist_sum` over `loss_mesh`:

https://github.com/pytorch/torchtitan/blob/main/torchtitan/trainer.py#L793-L802

```python
global_avg_loss, global_max_loss, global_ntokens_seen = (
    dist_utils.dist_sum(loss, loss_mesh),
    dist_utils.dist_max(local_avg_loss, loss_mesh),
    dist_utils.dist_sum(
        torch.tensor(
            self.ntokens_seen, dtype=torch.int64, device=self.device
        ),
        loss_mesh,
    ),
)
```

Where `loss_mesh` size = `dp_replicate x dp_shard x cp` (includes CP). Both CP ranks in the same DP group load the same full-size batch from the dataloader, so the same tokens are counted `cp_degree` times.

## Reproduction

Same model, same `global_batch_size=32`, same `seq_len=98304`:

| Config | grad_accum | per rank / step | dist_sum (loss_mesh, 8 ranks) | actual tokens / step | inflation |
|---|---|---|---|---|---|
| dp=8, cp=1 | 4 | 4 x 98304 = 393K | **3.15M** ✓ | 3.15M | 1x |
| dp=4, cp=2 | 8 | 8 x 98304 = 786K | **6.29M** ✗ | 3.15M | 2x |
| dp=2, cp=4 | 16 | 16 x 98304 = 1,573K | **12.58M** ✗ | 3.15M | 4x |

The inflation factor equals `cp_degree` exactly. When `dp` shrinks to make room for CP, `grad_accum` increases proportionally, and each CP rank counts the full sequence before the split.

On wandb, the CP=2 loss curve is stretched 2x and CP=4 is stretched 4x along the x-axis when plotted against `n_tokens_seen`, making it appear that CP needs more tokens to converge.

## NB

`tps` and `mfu` are **not affected** — they divide by `non_data_parallel_size` (which includes `cp`), correctly canceling the inflation. Only the cumulative `n_tokens_seen` counter lacks this normalization.

### Versions

- torchtitan: `main` branch (commit `9f055b4`)
- PyTorch: 2.11.0+cu128
- Reproduced on latest torchtitan main, the code path is unchanged

### Config (cp=2 example)

```toml
[training]
local_batch_size = 1
global_batch_size = 32
seq_len = 98304

[parallelism]
data_parallel_replicate_degree = 1
data_parallel_shard_degree = 4
tensor_parallel_degree = 1
pipeline_parallel_degree = 1
context_parallel_degree = 2
```

### PR DESCRIPTION
**Pull request by pianpwk**
Fixes https://github.com/pytorch/torchtitan/issues/2983

Before (DP=4, CP=1): `torchrun --nproc_per_node=4 --local-rank 0 -m torchtitan.train --module llama3 --config llama3_debugmodel_flex_attn --training.steps 5 --metrics.log_freq 1 --parallelism.data_parallel_shard_degree 4 --parallelism.context_parallel_degree 1 --training.global_batch_size 32 --training.seq_len 2048`
```
[titan] 2026-04-15 16:43:45,087 - root - INFO - [DEBUG] n_tokens_seen: 327680.0
[titan] 2026-04-15 16:43:45,087 - root - INFO - [31mstep:  5  [32mloss:  5.59510  [38;2;180;60;0mgrad_norm:  2.4778  [38;2;54;234;195mmemory:  1.09GiB(1.38%)  [34mtps: 104,455  [36mtflops: 7.48  [35mmfu: 2.40%[39m
```

Before (DP=2, CP=2): `torchrun --nproc_per_node=4 --local-rank 0 -m torchtitan.train --module llama3 --config llama3_debugmodel_flex_attn --training.steps 5 --metrics.log_freq 1 --parallelism.data_parallel_shard_degree 2 --parallelism.context_parallel_degree 2 --training.global_batch_size 32 --training.seq_len 2048`
```
[titan] 2026-04-15 16:37:32,439 - root - INFO - [DEBUG] n_tokens_seen: 655360.0
[titan] 2026-04-15 16:37:32,440 - root - INFO - [31mstep:  5  [32mloss:  5.63218  [38;2;180;60;0mgrad_norm:  2.4113  [38;2;54;234;195mmemory:  0.65GiB(0.82%)  [34mtps: 50,305  [36mtflops: 3.60  [35mmfu: 1.15%[39m
```

After (DP=4, CP=1):
```
[titan] 2026-04-15 16:44:35,135 - root - INFO - [DEBUG] n_tokens_seen: 327680.0
[titan] 2026-04-15 16:44:35,135 - root - INFO - [31mstep:  5  [32mloss:  5.60330  [38;2;180;60;0mgrad_norm:  2.2935  [38;2;54;234;195mmemory:  1.09GiB(1.38%)  [34mtps: 102,444  [36mtflops: 7.33  [35mmfu: 2.35%[39m
```

After (DP=2, CP=2):
```
[titan] 2026-04-15 16:45:17,312 - root - INFO - [DEBUG] n_tokens_seen: 327680.0
[titan] 2026-04-15 16:45:17,312 - root - INFO - [31mstep:  5  [32mloss:  5.58680  [38;2;180;60;0mgrad_norm:  2.2043  [38;2;54;234;195mmemory:  0.65GiB(0.82%)  [34mtps: 49,120  [36mtflops: 3.52  [35mmfu: 1.13%[39m
```

### CODE REVIEW

File: torchtitan/trainer.py [hunk 1]
```diff
@@ -540,7 +540,9 @@ def batch_generator(
                 raise DataloaderExhaustedError() from ex
             input_dict, labels = batch
             ntokens_batch = labels.numel()
-            self.ntokens_seen += ntokens_batch
```
Comment (tianyu-l) [inline torchtitan/trainer.py:543]: I think we should do it here
https://github.com/pytorch/torchtitan/blob/main/torchtitan/trainer.py#L646

where CP shards the inputs

**Review by tianyu-l — APPROVED** [review]
